todo REST API 기능 추가

tmxhsk99 opened the pull request:
안녕하세요
1주차 과제  ToDo REST API 과제 완성하여 
리뷰요청 드립니다.
감사합니다 : )

tmxhsk99 committed:
todo REST API 기능 추가

tmxhsk99 committed:
todo REST API 기능 추가

- 할일 추가 기능 추가
- 할일 수정기능 추가
- 할일 삭제 기능 추가
- 할일 조회 기능 추가
- 할일 상세 조회 기능 추가

hannut91 commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -1,10 +1,25 @@
+            InetSocketAddress address = new InetSocketAddress(DEFAULT_PORT);
+            HttpServer httpServer = HttpServer.create(address,0);
+            HttpHandler handler = new TaskHttpHandler();
+            httpServer.createContext(ROOT_PATH,handler);
```
```suggestion
            httpServer.createContext(ROOT_PATH, handler);
```

콤마 사이에 빈 칸이 없네요. 다른 코드에서는 빈 칸으로 구분하는 것 같아서 일관성있게 맞춰주면 좋을 것 같아요.
에디터의 자동 정렬 기능을 습관하시면 좋습니다. 혹은 파일을 저장할 때 마다 정렬되도록 할 수도 있습니다.

![image](https://github.com/CodeSoom/spring-week1-assignment-1/assets/14071105/e340498c-acc4-48aa-af26-123e7d24b838)

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,235 @@
+        }
+        if (isDeleteRequest()) {
+            deleteTask(exchange, path);
+        }
```
각 요청에 대해서 메서드가 처리하도록 해주셨네요. 만약에 실수로 Create요청에도 걸리고 Get요청에도 걸리면 의도하지 않은대로 실행될수도 있을 것 같아요.

무조건 이 중에 하나만 실행된다는 것을 표현하려면 else if를 사용하던지 혹은 return을 추가해서 표현할 수도 있을 것 같아요

hannut91 commented on app/src/main/java/com/codesoom/assignment/model/Task.java:
```diff
@@ -0,0 +1,45 @@
+                ", title='" + title + '\'' +
+                '}';
+    }
+}
```
코드 마지막 줄에는 빈 줄이 하나가 추가되어야 합니다. 그렇지 않으면 위와 같이 GitHub에서 표시가 됩니다.

![image](https://github.com/CodeSoom/spring-week1-assignment-1/assets/14071105/a2af9494-3c37-4671-b15b-125051d623ff)


이 것이 문제가 되는 이유는 POSIX 표준에서 Line을 `개행 문자로 끝나는 0개 이상의 개행 문자가 아닌 문자들의 시퀀스`으로 정의하고 있기 때문입니다.

```jsx
// Hello
```

만약 위와 같이 빈 줄이 없는 코드에서 아래에 코드를 추가하게 되면 Git은 다음과 같이 표시합니다.

```jsx
- // Hello
+ // Hello
+ // World
```

저는 단순히 아래 한 줄만 추가했는데 왜 세줄이 변경되었다고 나오는 걸까요? 그 이유는 줄바꿈 문자 `\n`가 들어갔기 때문입니다.

```jsx
- // Hello
+ // Hello\n
+ // World
```

만약 미리 빈 줄을 하나 추가한 상태에서 아래에 새로운 코드를 작성할 경우에는 다음과 같이 보여집니다.

```jsx
+ // World
```

이미 `\n`이 추가되어 있기 때문이죠. 따라서 코드 마지막 줄에 빈 줄 하나를 항상 추가해주는게 좋습니다.

혹은 자동으로 추가할 수 있도록 설정하는 방법도 있습니다. 종립님의 블로그 글을 참고해 주세요

See also

  - [파일-마지막에-자동으로-1개의-개행-문자-추가하기](https://johngrib.github.io/wiki/intellij/#%ED%8C%8C%EC%9D%BC-%EB%A7%88%EC%A7%80%EB%A7%89%EC%97%90-%EC%9E%90%EB%8F%99%EC%9C%BC%EB%A1%9C-1%EA%B0%9C%EC%9D%98-%EA%B0%9C%ED%96%89-%EB%AC%B8%EC%9E%90-%EC%B6%94%EA%B0%80%ED%95%98%EA%B8%B0)

hannut91 commented on app/src/main/java/com/codesoom/assignment/model/Task.java:
```diff
@@ -0,0 +1,45 @@
+                "id=" + id +
+                ", title='" + title + '\'' +
+                '}';
+    }
```
String.format을 사용해서 표현하는 건 어떨까요? 다음과 같이 표현할 수 있을거에요

```java
return String.format("{ id = %s, title = %s }", id, title);
```

더 좋은건 Template literal이지만 자바에서는 지원하질 않네요 ㅠ 만약 이런게 지원하는 코틀린을 사용했다면 다음과 같이 사용할 수 있을거에요

```kotlin
return "{ id = $id, title = $title }"
```

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,235 @@
+        } catch (NoSuchElementException e) {
+            e.printStackTrace();
+            exchange.sendResponseHeaders(HttpStatus.NOT_FOUND.getCode(), 0);
+        } finally {
+            exchange.close();
+        }
```
finally구문은 언뜻 보면 편리해 보이지만 사실 굉장히 복잡하고 모호하게 해석될 수 있는 여지가 많아요. 예를들어서

```java
public String test() {
    try {
        return "Hello";
    } finally {
        return "myname";
    }
}
```

위 메서드 실행 결과로 어떤 것이 반환될거라고 생각되시나요? 위 코드에서는 myname이 return됩니다. 이상하게 동작하죠? 저렇게 쓰지는 않겠지만 예상하기 어려울 때가 많습니다. 그래서 fianlly를 쓰지 않는 것이 좋습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,235 @@
+    /**
+     * 아이디에 해당하는 할일을 찾는다.
+     */
+    private void findByIdTask(HttpExchange exchange, String path) throws IOException {
```
여기서 모든 것을 처리하고 있어서 좀 복잡하게 느껴집니다. 적절한 책임으로 나눌 수 있을까요?

tmxhsk99 replied:
안녕하세요!
수정후 질문드립니다 혹시 책임을 나누실때 어떤 기준이 있으신가요?
적절한 책임으로 나누라고 책에도 자주 나오는데
아 이곳에는 책임이 몰려있구나 하게 알아채지 못하는 경우가 많고 
책임을 분리해도 이렇게 분리하는게 맞는가... 약간 불안감이 듭니다
혹시 어떤 기준으로 책임 구분하시는지 궁금합니다!

hannut91 replied:
어떤 책임이 있고 어떻게 나누어야 하는지 어려움이 있으시군요.
관심사의 분리라고도 하는데 관심사의 분리는 변경의 이유에 따라서 분리한다고 보시면 돼요.

```java
private Task updateTask(Task findTask, String title) throws NoSuchElementException {
    int findTaskIndex = taskList.indexOf(findTask);
    if (findTaskIndex == -1) {
        throw new NoSuchElementException("해당 할일이 존자하지 않습니다.");
    }

    findTask.setTitle(title);
    taskList.set(findTaskIndex, findTask);

    return findTask;
}
```

여기서 보면 이 코드가 변경이 될 수 있는 여지가 무엇이 있을까요?

```java
int findTaskIndex = taskList.indexOf(findTask);
```

할 일 목록을 List로 관리하고 있는데 만약 Map이나 다른 자료구조를 사용하게 된다면 여기에 코드에 변경이 일어나겠죠?

```java
if (findTaskIndex == -1) {
    throw new NoSuchElementException("해당 할일이 존자하지 않습니다.");
}
```

예외 메시지가 변경될때도 여기가 변경될 수 있겠네요.

```java
taskList.set(findTaskIndex, findTask);
```

여기도 마찬가지로 taskList가 변경되면 변경될 여지가 있어요.

이런식으로 코드에 변경이 될 이유가 한 가지만 있어야 한다는게 단일 책임 원칙이에요.

여기에 어떤 변경이 있을 수 있을까? 변경이 된다면 여기 코드는 어떻게 영향을 받을까? 변경을 여기 코드에서 격리시킬 수 있을까? 이런 고민을 해보면 좋을 것 같아요

tmxhsk99 replied:
각각 메서드로 분리하여 변경이 이루어질 경우 
해당 메서드에서만 변경이 1가지만 이루어지도록 수정하였습니다.
이렇게하면 메서드는 분리는 되었지만 메서드명(updateTaskInList,validateTaskIndex)이 리스트에게 종속적이게 이름이 지어져버리는것 같습니다.
처음부터 추상적으로 메서드이름을 짓는게 좋을까여 아니면 처음에 직관적이게 짓고 
변경되면 그때 메서드명도 변경하는게 좋을까여?

hannut91 replied:
메서드명이 구현에 의존하고 있네요. 마치 메서드가 하는 일을 구체적으로 작성해서 메서드가 하는 일이 조금이라도 바뀔 경우 메서드 이름도 같이 바뀔 가능성이 높아보입니다. 따라서 메서드 이름도 추상적으로 적는게 좋을 것 같습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,235 @@
+            Long id = parsePathToTaskId(path);
+            Task findTask = searchTask(id);
+            taskList.remove(findTask);
+            exchange.sendResponseHeaders(HttpStatus.NO_CONTENT.getCode(), 0);
```
enum을 사용해서 어떤 응답을 주는지 명시해 주셨네요. 성공을 주는 역할을 하는 객체를 만들어서 책임을 전가할 수도 있습니다.

```java
class ResponseSuccess {
    private static final int STATUS_CODE = 200;

    private final HttpExchange exchange;

    ResponseSuccess(HttpExchange exchange) {
        this.exchange = exchange;
    }

    public void send(String content) throw IOException {
        httpExchange.sendResponseHeaders(200, content.getBytes().length);

        OutputStream outputStream = httpExchange.getResponseBody();

        outputStream.write(content.getBytes());
        outputStream.close();
    }
}
```

tmxhsk99 replied:
이렇게 하면 훨씬 깔끔하군요!
그런데 보다보면 저는 또 모든 http status 코드를 넘겨서 
각각 해당하는 로직에 따라 나누어지고 싶어집니다
그렇게 보면 객체의 자율성이 낮아지는것(다른 객체가 하라는 것처럼 느껴짐...)이 아닌가 생각이 들기도하고
오버 엔지니어링이 아닌가 생각도 들고  
고민이 됩니다 어떻게 생각하시나요?

hannut91 replied:
다른 객체한테 시키다보니 지금 객체가 하고 있는 일이 별로 없는 것 같아서 자율성이 낮아진다고 느껴지시나봐요.
객체지향 프로그래밍은 자신이 해야할 일만 하고 나머지 일은 다른 객체한테 최대한 떠넘기는 프로그래밍이에요.
해당 객체가 하는 일이 자율성이 높아져서 오히려 더 자율적인 객체가 되니 더 나은 방향이라고 느껴져요.

tmxhsk99 committed:
fix : codeReformat 모드 적용 후 적용

tmxhsk99 committed:
fix : 로직상의 의도되지 않은 코드실행 방지

tmxhsk99 committed:
fix : POSIX Line 정의에 맞도록 수정

tmxhsk99 committed:
fix : String.format 형식으로 toString 수정

tmxhsk99 committed:
fix : finally 구문 삭제

tmxhsk99 committed:
fix : 기존의 findByIdTask메서드의 책임 분리

- getTask메서드는 해당 할일을 응답하고, findByIdTask는 할일을 찾는 책임으로 분리

tmxhsk99 committed:
fix : 성공을 주는 역할 객체로 수정

tmxhsk99 committed:
Resolved merge conflicts

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,239 @@
+    /**
+     * 요청 받은 할일을 저장 및 저장된 내용을 응답한다.
+     */
+    private void createTasksProcess(HttpExchange exchange, String body) throws IOException {
+        try (BufferedOutputStream bufferedOutputStream = new BufferedOutputStream(exchange.getResponseBody())) {
+            String task = createTask(body);
+            bufferedOutputStream.write(task.getBytes(StandardCharsets.UTF_8));
+            new ResponseSuccess(exchange).send(task);
+        } catch (JsonProcessingException e) {
+            e.printStackTrace();
+            exchange.sendResponseHeaders(HttpStatus.BAD_REQUEST.getCode(), 0);
+            exchange.close();
+        } catch (IOException e) {
+            exchange.close();
+            e.printStackTrace();
+            throw e;
+        }
+    }
```
너무 복잡해 보여요. 어떻게 하면 단순하게 코드를 읽을 수 있을까? 고민해 보시면 좋을 것 같아요. 이럴 때 좋은 방법은 그냥 이렇게 이야기하는 거에요.

`어떻게 구현할지는 모르겠지만 이랬으면 좋겠다`

그리고 코드를 그 모양대로 짜요.

```java
pirvate void createTaskProcess(HttpExchange exchange, String body) throws IOException {
    Body body = new Body(exchange);
    Task task = body.read(Task.class);
    Task createdTask = createTask(task);
    new ResponseSuccess(exchange).send(createdTask);
}
```

저라면 이런 모양이면 잘 이해할 수 있을 것 같아요. 상상으로 쓴거에요. 그다음에 여기에 맞춰서 코드를 작성해요.
물론 코드를 작성하다보면 잘못생각한것도 있고 더 좋은 방향도 떠오르게 될거에요. 그럼 그에 맞춰서 수정하면 돼요.

tmxhsk99 replied:
제가 작성하다보니 간과하게되는 면이 있는것같습니다;;(복잡해보인다고 생각을 아예 못함)
이후에는 해당방법 의식해서 적용해보도록 노력해보겠습니다!
꿀팁 감사합니다!

tmxhsk99 committed:
fix : taskUpdate 책임 분리

tmxhsk99 committed:
refactor : createTaskProcess 코드 리팩터링

tmxhsk99 committed:
fix : read 메서드 제너릭사용으로 변경

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,242 @@
+            taskList.add(task);
+            taskId++;
+            content = objectMapper.writeValueAsString(task);
+        } catch (JsonProcessingException e) {
+            e.printStackTrace();
+            throw e;
+        }
```
예외를 잡고나서 스택을 프린트하고 나서 다시 던지는 이유가 있을까요? 어디서 에러가 났었는지 확인하고 싶으셨던거죠?

tmxhsk99 replied:
아... 그 책임 분리할때 처음에 뭔가 잘못되었는 지 
기능이 동작을 안해서 에러를 찾으려고 찍었던것 같습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,242 @@
+            task.setId(taskId);
+            taskList.add(task);
+            taskId++;
+            content = objectMapper.writeValueAsString(task);
```
여기서 문자열로 만들어서 반환하는 이유는 응답을 줄 때 문자열로 전달해야되서 그런 것 같아요. 이 메서드는 할 일을 추가하는 일만 해야 하는데 응답을 위한 일도 하고 있는 것처럼 보이네요.

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,242 @@
+    /**
+     * 아이디에 해당하는 할일을 찾는다.
+     */
+    private Task findByIdTask(String path) {
```
할 일의 아이디로 할 일을 찾는 것이니 findByTaskId라는 이름이 좀 더 좋아 보이네요. 어떠세요? 그리고 이런 이름을 사용한다면 파라미터로 id를 받을 것 같은데 path를 받네요. 사실 이 메서드의 정확한 이름은
parsePathAndFindByTaskId가 되어야 하죠. 두 가지의 일을 하고 있으니까요.
여기서 path로부터 id를 읽는 것은 좋아보이지 않네요.

tmxhsk99 replied:
parsePathAndFindByTaskId 같은경우 2가지 일 하니까
그냥 분리하는게 좋아보이는데 어떻게 생각하세여? 
메서드이름도 길어지니 한눈에도 확 안들어오는 느낌이들어서...
각각 findByTaskId(Task task), parsePathToTaskId(String path) 요렇게 분리하니까
searchTask는 그냥 호출만하고 아무내용이 없어서 searchTask를 대체하였습니다.

hannut91 replied:
네 ㅎㅎ 코드 확인했습니다. 각각 하는 일로 분리하는게 좋아 보입니다. id를 얻는 방법이 달라질 경우 parsePathToTaskId이 변경될 것이고, findByTaskId를 id가 주어지면 자신이 해야할 일을 하니까요

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,242 @@
+    private void changeTask(Task findTask, String title, int findTaskIndex) {
+        findTask.setTitle(title);
+        updateTaskInList(findTaskIndex, findTask);
+    }
```
findTask라는 이름은 Task를 찾아! 라고 명령하는 것 같아요. 그래서 메서드 이름처럼 보이네요.
changeTask의 입장에서 바라보면 주어진 할 일이 찾아서 얻은 할 일인지 아니면 다른 일을 하다가 할 일을 수정하게 되었는지는 관심이 없어요. 주어진게 할 일이라는 사실 뿐이죠. `할 일을 찾아서 수정한다`라는 문맥이 이 메서드로 침투한거에요.
따라서 그냥 task라는 이름이 더 괜찮아 보여요

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,242 @@
+
+    }
+
+    private int findTaskElement(Task task) {
+        int index = taskList.indexOf(task);
+        validateTaskIndex(index);
+        return index;
+    }
+
+    private void validateTaskIndex(int index) {
+        if (index == -1) {
+            throw new NoSuchElementException("해당 할일이 존재하지 않습니다.");
+        }
+    }
```
메서드로 분리해서 한 가지 일만 하도록 하고 표현력을 높이려고 시도하셨네요. 메서드로 동작을 추상화하면 메서드의 변경에 대해서 영향을 적게 받을 수 있고 독립적으로 발전할 수 있어서 메서드로 분리하는 것이 좋을 수 있어요. 하지만 메서드가 기능을 숨기기 때문에 더 이해하기 어려워질 수있어요. 이 경우가 메서드로 분리해서 오히려 어려워진 경우에요.

인덱스가 `-1` 이라는 것은 인덱스의 범위가 벗어났다는 것이고 이것은 `indexOf` 의 결과로 존재하지 않는다는 것을 바로 알 수 있어요. 하지만 validate라는 메서드를 사용하면 index가 -1인지, 아니면 주어진 범위를 벗어난건지 혹은 숫자가 아니라 영문이 왔는지 여러가지로 해석될 수 있어요. 따라서 이 메서드 추출은 오히려 이해하기 어렵게 만들어요.

tmxhsk99 replied:
먼가 하면서도 찜찜했는데 그렇군요!
메서드로 분리할 때는 좀더 적확한 메서드명을 짓도록 신경써야겠습니다

tmxhsk99 committed:
fix : 불필요한 printStackTrace 제거

tmxhsk99 committed:
fix : jsonString 파싱책임 분리

tmxhsk99 committed:
fix : findByIdTask 관련 책임 분리및 메서드 명 변경

- findByTaskId로 메서드명 변경
- path를 파싱하는 로직 따로 분리
- searchTask 에게 위임만하여 searchTask 로직 findBayTaskId로 로직이동
  기존 searchTask 메서드 삭제

tmxhsk99 committed:
refactor : validateTask 메서드명 수정

- AS-IS: validateTask , TO-BE: checkIfTaskExists

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,240 @@
+
+        settingRequestMessage(exchange);
+
+        if (isCreateRequest()) {
+            createTasksProcess(exchange);
+        } else if (isGetListRequest()) {
+            getTasksProcess(exchange);
+        } else if (isGetRequest()) {
+            getTask(exchange, path);
+        } else if (isUpdateRequest()) {
+            updateTaskProcess(exchange, path);
+        } else if (isDeleteRequest()) {
+            deleteTask(exchange, path);
+        }
```
```java
class CreateTaskHandler {
  private String method;
  private String path;
  private HttpExchange exchange;
  
  public CreateTaskHandler(HttpExchange exchange) {
    this.exchange = exchange;
    this.method = "POST";
    this.path = "/tasks";
  }

  public boolean isRequest() {
    URI uri = exchange.getRequestURI();
    return this.method == "POST" && this.path.equals(uri.getPath());
  }

  public void handle() {
    // Do something
  }
}
```
위와 같은 객체를 만들어서
```java
CreateTaskHandler createTaskHandler = new CreateTaskHander(exchange);
if (createTaskHandler.isRequest()) {
  createTaskHandler.handle();
}
```

이런식으로 하면 어떨까 저도 한 번 고민해보았습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,240 @@
+
+    }
+
+    private int findTaskElement(Task task) throws NoSuchElementException {
+        int index = taskList.indexOf(task);
+        checkIfTaskExists(index);
+        return index;
+    }
```
할 일을 찾은 후에 인덱스를 반환하는군요. 이것도 TaskList가 List로 구현되어 있기 때문에 이렇게 구현되어있는 것으로 보입니다. 만약에 TaskList를 Map에다 관리한다고 하면 어떻게 될까요? 이 index에 의존하고 있는 모든 메서드들이 영향을 받겠죠?

tmxhsk99 replied:
TaskRepository 인터페이스를 정의하고 구현체인 TaskListRepository 로 분리하였습니다.
종속적인 메서드는 구현체인 TaskListRepository 에만 존재하고 공통적인 스펙은 TaskRespository 인터페이스에 정의하였습니다.
update 로직은 기존의 ArrayList에 종속적인 메서드는 삭제하고
찾은객체를 그냥 수정하는 방식으로 간단히 바꾸었습니다.

tmxhsk99 committed:
refactor : handler를 계층구조로 변경, repository 인터페이스 추가

tmxhsk99 committed:
refactor : TsakHttpHandler 분기로직 Map으로 수정, update로직수정

- handlerMap을 통해 분기하돌록 수정
- update 로직수정
	- AS-IS: ArrayList에인덱스가 존재하는지 확인후 해당 인덱스 요소
	  교체
	- TO-BE: 찾은 객체를 수정하도록 수정

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/TaskHandler.java:
```diff
@@ -0,0 +1,33 @@
+    public abstract boolean isRequest();
+
+    protected Long parsePathToTaskId(String path) {
+        String[] splitUrl = path.split("/");
+        Long id = Long.parseLong(splitUrl[splitUrl.length - 1]);
```
경로 마지막에 있는 값을 id로 읽고 있군요. 정규식을 이용한 방법도 있겠네요

tmxhsk99 replied:
앗 정규식으로 하는 방법도 있군요... (처음 앎)
다음부터는 적극 활용하도록 하겠습니다!

hannut91 commented on app/src/main/java/com/codesoom/assignment/handler/TaskHttpHandler.java:
```diff
@@ -0,0 +1,39 @@
+    public void handle(HttpExchange exchange) throws RuntimeException {
+        initHandler(exchange);
+
+        handlerMap.entrySet().stream()
+                .filter(entry -> entry.getValue().isRequest())
+                .findFirst()
+                .ifPresent(entry -> {
+                    try {
+                        entry.getValue().handle();
+                    } catch (IOException e) {
+                        throw new RuntimeException(e);
+                    }
+                });
```
Optional의 값이 있는 경우를 하려고 하신 것 같습니다. 이런 경우는 안에서 다하는 것보다는 적절히 구분해주는게 좋아보이네요

```suggestion
        var entry = handlerMap.entrySet().stream()
                .filter(entry -> entry.getValue().isRequest())
                .findFirst();
       if (entry.isPresent()) {
           entry.getValue().handle();
       }
```

그리고 계속 예외를 잡고 다시 던지기 방법을 사용하고 계신 것 같은데, 특별한 이유가 없다면 에러는 한 번 만 잡는게 좋고, 잡아야 한다면 이유가 있어야 합니다. 여기서 단순히 예외를 바꿔서 던지는 것은 별로 좋아보이진 않아요

tmxhsk99 replied:
람다식 내부에서 억지로 전부 끝내려고하다보니 저렇게 된것같습니다.
람다식 밖에서 throw로 던져서 하나의 예외형태는 한군데서 처리하는게 기본적으로는 좋다는 말씀이시군요.
궁금한 점이 있습니다
에러를 중간에서 잡는다고한다면 어떤경우가 있을까요?
일단 제가 이해한것은 만약에 어떤 예외가 발생할 경우 뭔가 추가적으로
처리할 로직이 있는경우 중간에 잡아 주어야 한다는 걸까요?

hannut91 replied:
에러를 중간에서 잡는 경우는 사실 별로 없습니다. 더 큰 맥락을 알아야 에러를 처리할 수 있기 떄문이죠. 그래서 `에러를 한 번만 잡아라!` 라는 조언도 있습니다. 하지만 `에러를 잡고 다시 던지기`라는 것도 있습니다. 이것은 에러를 잡아서 다시 무언가를 바꿔서 던질려고 하는겁니다. 

예를들어 지금 작성해주신 코드`IOException`를 `RuntimeException`를 바꿔서 던진게 에러를 잡고 다시 던진거죠. 대부분 이런 경우는 에러 객체를 변경해주기 위해서 씁니다. 그런데 작성해주신 것을 보면 커스텀 에러 객체가 아니라 RuntimeException을 사용하고 있어서 다시 던지는 이유가 RuntimeException으로 바꾸는 이유인데 그게 납득이 되지는 않죠. 예를들어서 IOException 예외를 잡아서 EntryIsNotDefinedException으로 던진다면 의미를 더 부여하기 위해서 하는 것으로 볼 수 있겠죠

> 문제 발생 사실을 전달해야 하는 방향은 함수 호출의 역방향이다. 문제에 따라서는 더 큰 맥락을 알아야 문제를 해결하거나 대안을 제시할 수 있는데, 호출 사슬의 아래쪽보다 위쪽에 있는 함수들이 더 큰 맥락을 알고 있기 때문이다. - 핵심 코틀린 프로그래밍 - 5장 예외

hannut91 reviewed (approved):
이번 한 주간 정말 여러가지 시도를 했네요. 지금은 배우는 과정이니 무엇이든 시도해봐도 괜찮지만, 실무에서는 너무 복잡하게 문제를 해결하고 있지 않나 고민을 많이 해야됩니다.

고생 많으셨어요

tmxhsk99 commented:
윤석님도 한주동안 고생많으셨습니다 ㅠㅠ( 부족한 코드 봐주시느라... )
항상 뭔가 베스트 프렉티스나 기본이 부족한 면이 있어서 최대한 말씀해주시는거
기억하면서 앞으로도 해보겠습니다
감사합니다!

tmxhsk99 committed:
refactor : taskHandler 에러처리 로직및 path 파싱로직 수정
